File: apps/mobile/targets/credentials-provider/SharedVaultStore.swift
import Foundation
import Security

/// The two things this extension reads across the process boundary from the
/// main app — never anything else. Both are written by the main app: the
/// VMK via react-native-keychain (apps/mobile/src/lib/biometrics.ts, with
/// its accessGroup pointed at SHARED_ACCESS_GROUP so this target can read
/// the same Keychain item), the ciphertext cache via a small native module
/// call from App.tsx whenever the decrypted `items` list changes (still
/// ciphertext at rest here — the exact same wrapped_item_key/content shape
/// vault_items already has server-side, no new plaintext exposure).
enum SharedVaultStoreError: Error {
    case keychainUnavailable(OSStatus)
    case cacheMissing
    case cacheUnreadable
}

/// Must match apps/mobile/app.json's `ios.entitlements` exactly (both the
/// main app's and this target's expo-target.config.js mirror the same
/// value) — see build plan §7 step 7 addendum on why an App Group +
/// Keychain Access Group are the only channel here, not a live IPC relay.
/// The RN side writes the cache via react-native-shared-group-preferences
/// (apps/mobile/src/lib/autofillSync.ts), which stores it as a string value
/// in this same App Group's UserDefaults suite — that's why this reads
/// UserDefaults rather than a raw file in the shared container.
enum SharedVaultConfig {
    static let appGroupID = "group.com.yourorg.passwordmanager.shared"
    static let keychainAccessGroup = "com.yourorg.passwordmanager.shared" // Xcode prefixes $(AppIdentifierPrefix) automatically for entitlement matching
    static let keychainService = "com.yourorg.passwordmanager.quick-unlock"
    static let keychainAccount = "vault"
    static let cacheDefaultsKey = "vault_items_cache"
}

/// Mirrors the (small, ciphertext-only) projection apps/mobile writes —
/// deliberately not the full vault_items row shape, just what autofill
/// needs. See App.tsx's cache-writing effect.
struct CachedWrappedPayload: Codable {
    let nonce: String
    let ciphertext: String
}

struct CachedItemCiphertext: Codable {
    let nonce: String
    let ciphertext: String
    let aad: String
}

struct CachedVaultItem: Codable {
    let id: String
    let type: String
    let wrappedItemKey: CachedWrappedPayload
    let content: CachedItemCiphertext
}

enum SharedVaultStore {
    /// Reading a BIOMETRY_CURRENT_SET-protected Keychain item triggers the
    /// system Face ID/Touch ID sheet automatically — no custom biometric UI
    /// here, matching mobile design plan §3's "always defer to the OS
    /// sheet" rule (same posture as apps/mobile/src/lib/biometrics.ts).
    static func readVmkBase64() throws -> String {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: SharedVaultConfig.keychainService,
            kSecAttrAccount as String: SharedVaultConfig.keychainAccount,
            kSecAttrAccessGroup as String: SharedVaultConfig.keychainAccessGroup,
            kSecReturnData as String: true,
            kSecUseOperationPrompt as String: "Unlock your vault to autofill",
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess, let data = item as? Data, let vmkBase64 = String(data: data, encoding: .utf8) else {
            throw SharedVaultStoreError.keychainUnavailable(status)
        }
        return vmkBase64
    }

    static func readCachedItems() throws -> [CachedVaultItem] {
        guard let defaults = UserDefaults(suiteName: SharedVaultConfig.appGroupID) else {
            throw SharedVaultStoreError.cacheMissing
        }
        guard let json = defaults.string(forKey: SharedVaultConfig.cacheDefaultsKey),
            let data = json.data(using: .utf8)
        else {
            throw SharedVaultStoreError.cacheMissing
        }
        guard let items = try? JSONDecoder().decode([CachedVaultItem].self, from: data) else {
            throw SharedVaultStoreError.cacheUnreadable
        }
        return items
    }
}

File: apps/mobile/native/android/autofill/SharedVaultStore.kt
package com.yourorg.passwordmanager.autofill

import android.content.Context
import org.json.JSONArray

/**
 * Reads the ciphertext item cache apps/mobile/src/lib/autofillSync.ts writes
 * via react-native-shared-group-preferences (`useAndroidSharedPreferences:
 * true` — a plain private SharedPreferences file named after the "app
 * group" string, same package/process so no real cross-app sharing is
 * needed on Android, unlike iOS). Still ciphertext at rest — the same
 * wrapped_item_key/content shape vault_items has server-side.
 */
data class CachedWrappedPayload(val nonce: String, val ciphertext: String)
data class CachedItemCiphertext(val nonce: String, val ciphertext: String, val aad: String)
data class CachedVaultItem(
    val id: String,
    val type: String,
    val wrappedItemKey: CachedWrappedPayload,
    val content: CachedItemCiphertext,
)

object SharedVaultStore {
    // Must match apps/mobile/src/lib/autofillSync.ts's APP_GROUP/CACHE_KEY exactly.
    private const val APP_GROUP = "group.com.yourorg.passwordmanager.shared"
    private const val CACHE_KEY = "vault_items_cache"

    fun readCachedItems(context: Context): List<CachedVaultItem> {
        val json = context.getSharedPreferences(APP_GROUP, Context.MODE_PRIVATE)
            .getString(CACHE_KEY, null) ?: return emptyList()
        return try {
            val array = JSONArray(json)
            (0 until array.length()).mapNotNull { i ->
                val row = array.optJSONObject(i) ?: return@mapNotNull null
                val wrappedKey = row.optJSONObject("wrappedItemKey") ?: return@mapNotNull null
                val content = row.optJSONObject("content") ?: return@mapNotNull null
                CachedVaultItem(
                    id = row.optString("id"),
                    type = row.optString("type"),
                    wrappedItemKey = CachedWrappedPayload(
                        nonce = wrappedKey.optString("nonce"),
                        ciphertext = wrappedKey.optString("ciphertext"),
                    ),
                    content = CachedItemCiphertext(
                        nonce = content.optString("nonce"),
                        ciphertext = content.optString("ciphertext"),
                        aad = content.optString("aad"),
                    ),
                )
            }
        } catch (_: Exception) {
            emptyList()
        }
    }
}
